GS/SW: Mask color gradients to prevent incorrect clamping.

TJnotJT opened the pull request:
### Description of Changes
Mask color gradients before converting 32->16 bit in prim setup to prevent unwanted clamping.

### Rationale behind Changes
When gradients are too large to fit in the fix point format, rolling them over will still preserve the correct colors in the scanline renderer. This makes sure the roll over happens correctly to prevent graphical bugs.

Fixes https://github.com/PCSX2/pcsx2/issues/6459
Fixes https://github.com/PCSX2/pcsx2/issues/10210.

### Suggested Testing Steps
Testing any games with the SW renderer on both SSE and AVX2 builds.

### Did you use AI to help find, test, or implement this issue or feature?
Looking up aarch64 instructions.

### Credits
Co-authored-by: TellowKrinkle

TJnotJT committed:
GS/SW: Mask color gradients to prevent incorrect clamping.

Co-authored-by: TellowKrinkle

TellowKrinkle commented:
Didn't look super hard but I think you're missing save & restore of xmm15 on windows.  Yay for calling conventions being fun.

TJnotJT commented:
> Didn't look super hard but I think you're missing save & restore of xmm15 on windows. Yay for calling conventions being fun.

Good point, I had forgotten about this. Just fixed it.

TellowKrinkle commented on pcsx2/GS/Renderers/SW/GSSetupPrimCodeGenerator.arm64.cpp:
```diff
@@ -225,13 +225,16 @@ void GSSetupPrimCodeGenerator::Color()
 		armAsm->Rev64(_vscratch.V4S(), v2.V4S());
 		armAsm->Uzp1(v2.V4S(), v2.V4S(), _vscratch.V4S());
-		armAsm->Sqxtn(v2.V4H(), v2.V4S());
+		armAsm->Uqxtn(v2.V4H(), v2.V4S());
```
```suggestion
		armAsm->Uzp1(v2.V8H(), v2.V8H(), v2.V8H());
```
Can drop the dup afterwards too

TellowKrinkle reviewed (commented):
arm has non-saturating pack instructions, so no separate masking is needed

GH doesn't let me suggest edits on existing code, but the entire second section can be replaced with
```cpp
			// VectorI r = VectorI(dr * shift[1 + i]);

			armAsm->Fmul(v2.V4S(), v0.V4S(), VRegister(4 + i, kFormat4S));
			armAsm->Fcvtzs(v2.V4S(), v2.V4S());

			// VectorI b = VectorI(db * shift[1 + i]);

			armAsm->Fmul(v3.V4S(), v1.V4S(), VRegister(4 + i, kFormat4S));
			armAsm->Fcvtzs(v3.V4S(), v3.V4S());

			// m_local.d[i].rb = r.trn1_16(b); // Yeah I know this isn't in GSVector since that's mainly targeting x86 for now
			armAsm->Trn1(v2.V8H(), v2.V8H(), v3.V8H());
			armAsm->Str(v2, _local(d[i].rb));
```

TJnotJT committed:
GS/SW: Use non-saturating ARM instructions for color gradient setup.

This is more efficient on ARM, though the equivalent instructions are not currently used in the x64 JIT and C++ versions of GSVector.

Co-authored-by: TellowKrinkle

TellowKrinkle commented on pcsx2/GS/Renderers/SW/GSSetupPrimCodeGenerator.arm64.cpp:
```diff
@@ -225,14 +225,14 @@ void GSSetupPrimCodeGenerator::Color()
+		// local.d4.c = tmp.uzp1_16(tmp); // Yeah I know this isn't in GSVector since that's mainly targeting x86 for now
 		armAsm->Fmul(v2.V4S(), v16.V4S(), v3.V4S());
 		armAsm->Fcvtzs(v2.V4S(), v2.V4S());
+		armAsm->And(v2.V4S(), v17.V4S());
```
You missed this one

TellowKrinkle reviewed (approved):
Looks good assuming nothing breaks in the dump run

JordanTheToaster reviewed (approved):
Tested a variety of games and all seems fine on Windows with AVX2.

lightningterror reviewed (approved):
Did an sse4 dump and didn't notice any issues (700+ dumps showed diffs with no visual difference so maybe something could still slip).
